### PR TITLE
Fix: improve Ansible documentation

### PR DESCRIPTION
**Pull request by oxedions**
Ansible part improvement of the documentation.

### CODE REVIEW

File: resources/documentation/learn_ansible.rst [hunk 1]
```diff
@@ -691,5 +691,395 @@ Now regarding to Jinja2:
...
+Will display "lenneth".
+
+We are going to uses register and debug module to learn generic tasks actions.
+
+Loops
+^^^^^
+
+It is possible to make modules iterate. All possibilities are available here in `loop documentation <https://docs.ansible.com/ansible/latest/user_guide/playbooks_loops.html>`_ .
+
+BlueBanquise rely on two methods: *with_items* and *loop*. *with_items* is considered replaced by the Ansible team, but since *loop*, the replacement, does not cover all the abilities of *with_items*, it is still widely used.
```
Comment (actatux) [inline resources/documentation/learn_ansible.rst:792]: By `with_items`, do you mean `with_<lookup>`?
Otherwise, for my knowledge, can you show me examples where `with_items` cannot be replaced by `loop` and the `flatten` filter?

File: resources/documentation/learn_ansible.rst [hunk 2]
```diff
@@ -691,5 +691,395 @@ Now regarding to Jinja2:
...
+  - name: My message module
+    debug:
+      msg: "Values: {{ item }}"
+    loop:
+       - Blue
+       - Red
+       - Green
+       - Yellow
+
+This will execute this module 4 times, with each time an element of the list gives to loop. When using loops in ansible, the variable **item** store the value of the current loop index.
```
Comment (actatux) [inline resources/documentation/learn_ansible.rst:807]: s/gives/given/ ?
s/store/stores/

File: resources/documentation/learn_ansible.rst [hunk 3]
```diff
@@ -691,5 +691,395 @@ Now regarding to Jinja2:
...
+
+It is also possible to provide loop a list from the inventory, like the one we made before:
```
Comment (actatux) [inline resources/documentation/learn_ansible.rst:809]: ... provide the loop with a list ... like the one we wrote before

File: resources/documentation/learn_ansible.rst [hunk 4]
```diff
@@ -691,5 +691,395 @@ Now regarding to Jinja2:
...
+
+.. code-block:: yaml
+
+  - name: My message module
+    debug:
+      msg: "Values: {{ item }}"
+    loop: "{{ my_ship.equipment.sidekicks }}"
+
+Loop action can accept advanced filters or patterns. Have a look in the official documentation.
```
Comment (actatux) [inline resources/documentation/learn_ansible.rst:818]: rephrase?
_Refer to the Ansible documentation_

File: resources/documentation/learn_ansible.rst [hunk 5]
```diff
@@ -691,5 +691,395 @@ Now regarding to Jinja2:
...
+
+with_items works the same way:
+
+.. code-block:: yaml
+
+  - name: My message module
+    debug:
+      msg: "Values: {{ item }}"
+    with_items:
```
Comment (actatux) [inline resources/documentation/learn_ansible.rst:827]: `with_list` ? (preferred method to iterate over a list before `loop`)

File: resources/documentation/learn_ansible.rst [hunk 6]
```diff
@@ -691,5 +691,395 @@ Now regarding to Jinja2:
...
+       - Yellow
+
+  - name: My message module
+    debug:
+      msg: "Values: {{ item }}"
+    with_items: "{{ my_ship.equipment.sidekicks }}"
+
+So why not using loop all the time?
+
+Issue is when you need to use a conditional (see below) and that condition is used to check if the variable is defined.
```
Comment (actatux) [inline resources/documentation/learn_ansible.rst:840]: not clear to me

File: resources/documentation/learn_ansible.rst [hunk 7]
```diff
@@ -691,5 +691,395 @@ Now regarding to Jinja2:
...
+with_items will execute after the when, while loop execute before, and so could fail.
```
Comment (actatux) [inline resources/documentation/learn_ansible.rst:841]: are you sure?

File: resources/documentation/learn_ansible.rst [hunk 8]
```diff
@@ -691,5 +691,395 @@ Now regarding to Jinja2:
...
+.. code-block:: text
+
+  TASK [debug : Display content of the register] **********************************************************************************************************************************************************************************************
+  skipping: [management1] => (item=freya)
+  ok: [management1] => (item=lenneth) =>
+    msg: Profile
+  skipping: [management1] => (item=odin)
+  skipping: [management1] => (item=loki)
+
+So it acted as an AND.
```
Comment (actatux) [inline resources/documentation/learn_ansible.rst:961]: as a logical AND ?

File: resources/documentation/learn_ansible.rst [hunk 9]
```diff
@@ -691,5 +691,395 @@ Now regarding to Jinja2:
...
+
+.. code-block:: yaml
+
+  - name: There was a change
+    debug:
+      msg: "THE FILE WAS CHANGED!! HOW DARE YOU!!"
+
+Note that the name of this task (There was a change) is important and will be targeted by the notify.
+
+Create also a template, called our_file.j2 in templates folder, with the following content:
```
Comment (actatux) [inline resources/documentation/learn_ansible.rst:1005]: in **templates** folder (like **handlers** above)

File: resources/documentation/learn_ansible.rst [hunk 10]
```diff
@@ -691,5 +691,395 @@ Now regarding to Jinja2:
...
+
+.. code-block:: text
+
+  {{ my_value }}
+
+Now, in the main task, use this code:
+
+.. code-block:: yaml
+
+  - name: Display content of the register
```
Comment (actatux) [inline resources/documentation/learn_ansible.rst:1015]: there is no register

File: resources/documentation/learn_ansible.rst [hunk 11]
```diff
@@ -691,5 +691,395 @@ Now regarding to Jinja2:
...
+  - name: Template /tmp/out
+    template:
+      src: "our_file.j2"
+      dest: /tmp/out
+      owner: root
+      group: root
+      mode: 0644
+    notify: There was a change
+
+  - name: Display content of the register
```
Comment (actatux) [inline resources/documentation/learn_ansible.rst:1028]: there is no register

File: resources/documentation/learn_ansible.rst [hunk 12]
```diff
@@ -43,7 +43,7 @@ Also set, in this same file /etc/ansible/ansible.cfg, the roles_path value to /e
 
   roles_path    = /etc/ansible/roles
 
-And also change hash_behaviour to merge, as this is the most interesting algorithm:
+And also change hash_behaviour to merge, as this is the most interesting and mandatory algorithm for us:
```
[the remark this answers is not in the document]
Comment (oxedions) [reply resources/documentation/learn_ansible.rst:46]: Yes, I am working on this. I need to test the whole stack considering this, and update documentation :-(

File: resources/documentation/learn_ansible.rst [hunk 13]
```diff
@@ -691,5 +691,395 @@ Now regarding to Jinja2:
 You can experiment with this template to understand the whole mechanism. The whole Jinja2 documentation can be found here: `Jinja2 template designer <https://jinja.palletsprojects.com/en/2.10.x/templates/>`_
 
 You should now understand the very basis of Ansible.
+
+Time to investigate tasks advanced elements.
+
+Task again
+----------
+
+As discussed before, all task available modules can be found in the `official documentation <https://docs.ansible.com/ansible/latest/modules/modules_by_category.html>`_ .
```
[the remark this answers is not in the document]
Comment (oxedions) [reply resources/documentation/learn_ansible.rst:700]: Agreed.

File: resources/documentation/learn_ansible.rst [hunk 14]
```diff
@@ -691,5 +691,395 @@ Now regarding to Jinja2:
...
+    debug:
+      msg: "Hello world ! I am the debug module. You will use me a lot."
+
+You can also display variables this way:
+
+.. code-block:: yaml
+
+  - name: My message module
+    debug:
+      msg: "Hello world ! I am the hosts {{ inventory_hostname }}"
```
[the remark this answers is not in the document]
Comment (oxedions) [reply resources/documentation/learn_ansible.rst:723]: Yes

File: resources/documentation/learn_ansible.rst [hunk 15]
```diff
@@ -691,5 +691,395 @@ Now regarding to Jinja2:
...
+      failed: false
+      rc: 0
+      start: '2020-04-17 14:46:04.474423'
+      stderr: ''
+      stderr_lines: []
+      stdout: lenneth
+      stdout_lines:
+      - lenneth
+
+So, you can use the register to get status (changed: true, so something happens), to get return code (rc at 0 here), stderr and stdout, if it failed, etc.
```
[the remark this answers is not in the document]
Comment (oxedions) [reply resources/documentation/learn_ansible.rst:773]: I don't get it. you mean 'rc at 0 here' -> rc = 0 here ?

File: resources/documentation/learn_ansible.rst [hunk 16]
```diff
@@ -691,5 +691,395 @@ Now regarding to Jinja2:
...
+
+.. code-block:: yaml
+
+  - name: My message module
+    debug:
+      msg: "{{ my_result.stdout }}"
+
+Will display "lenneth".
+
+We are going to uses register and debug module to learn generic tasks actions.
```
[the remark this answers is not in the document]
Comment (oxedions) [reply resources/documentation/learn_ansible.rst:785]: Yes

File: resources/documentation/learn_ansible.rst (hunk 7 above)
Comment (oxedions) [reply resources/documentation/learn_ansible.rst:841]: Try it. This was one of the main issue I had when migrating to loop, when using ansible 2.7.

File: resources/documentation/learn_ansible.rst (hunk 8 above)
Comment (oxedions) [reply resources/documentation/learn_ansible.rst:961]: yes

File: resources/documentation/learn_ansible.rst (hunk 9 above)
Comment (oxedions) [reply resources/documentation/learn_ansible.rst:1005]: yes

File: resources/documentation/learn_ansible.rst (hunk 15 above)
[the remark this answers is not in the document]
Comment (actatux) [reply resources/documentation/learn_ansible.rst:773]: I would rephrase 'rc at 0' -> 'rc is 0' (or 'rc = 0')
Not sure 'at' is correct here.

File: resources/documentation/learn_ansible.rst (hunk 11 above)
Comment (oxedions) [reply resources/documentation/learn_ansible.rst:1028]: yes, good catch

File: resources/documentation/learn_ansible.rst (hunk 1 above)
Comment (oxedions) [reply resources/documentation/learn_ansible.rst:792]: I think it was when combined with when condition
If you use when to check if a variable is defined, and the variable is not defined, then with with_item it will be ok as when is checked before with item, while with loop, loop execute first. You need to provide a default to solve this.

Comment (oxedions) [reply resources/documentation/learn_ansible.rst:792]: And bluebanquise only uses with_items :-)

Comment (oxedions) [reply resources/documentation/learn_ansible.rst:792]: No other with_X

File: resources/documentation/learn_ansible.rst (hunk 2 above)
Comment (oxedions) [reply resources/documentation/learn_ansible.rst:807]: yes

File: resources/documentation/learn_ansible.rst (hunk 3 above)
Comment (oxedions) [reply resources/documentation/learn_ansible.rst:809]: ok

File: resources/documentation/learn_ansible.rst (hunk 4 above)
Comment (oxedions) [reply resources/documentation/learn_ansible.rst:818]: done

File: resources/documentation/learn_ansible.rst (hunk 5 above)
Comment (oxedions) [reply resources/documentation/learn_ansible.rst:827]: Since only with_item is used in BB, I wanted to focus on it first. Users can get the other with_lookup from documentation :-)

File: resources/documentation/learn_ansible.rst (hunk 6 above)
Comment (oxedions) [reply resources/documentation/learn_ansible.rst:840]: removed

File: resources/documentation/learn_ansible.rst (hunk 7 above)
Comment (oxedions) [reply resources/documentation/learn_ansible.rst:841]: removed

**Comment by oxedions** [comment]
Updated the file.
